Improve logLevel command

jiacheliu3 opened the pull request:
There are multiple improvements in this PR:
1. Make `logLevel` command work in HA. Previously in HA setup, the `logLevel` command will only look for a master process local to where the command is run. Now the command will inquire for the primary master.
2. Improve the output in various ways providing more informative printouts.

Below is an example of HA with master and job master being on different nodes.
```
$ hostname
AlluxioSandbox-log1-masters-1
$ bin/alluxio logLevel --logName=alluxio.master.job.plan --level=DEBUG
Primary job master: **AlluxioSandbox-log1-masters-1**:20002[job_master]
Job worker: AlluxioSandbox-log1-workers-2:30003[job_worker]
Job worker: AlluxioSandbox-log1-workers-1:30003[job_worker]
Worker: AlluxioSandbox-log1-workers-2:30000[worker]
Worker: AlluxioSandbox-log1-workers-1:30000[worker]
Primary master: **AlluxioSandbox-log1-masters-2**:19999[master]
AlluxioSandbox-log1-masters-1:20002[job_master]LogInfo{DEBUG, alluxio.master.job.plan, Setting Level to DEBUG}
AlluxioSandbox-log1-workers-2:30003[job_worker]LogInfo{DEBUG, alluxio.master.job.plan, Setting Level to DEBUG}
AlluxioSandbox-log1-workers-1:30003[job_worker]LogInfo{DEBUG, alluxio.master.job.plan, Setting Level to DEBUG}
AlluxioSandbox-log1-workers-2:30000[worker]LogInfo{DEBUG, alluxio.master.job.plan, Setting Level to DEBUG}
AlluxioSandbox-log1-workers-1:30000[worker]LogInfo{DEBUG, alluxio.master.job.plan, Setting Level to DEBUG}
AlluxioSandbox-log1-masters-2:19999[master]LogInfo{DEBUG, alluxio.master.job.plan, Setting Level to DEBUG}
```

Verified with Embedded and ZK setups.

jiacheliu3 commented:
@LuQQiu thanks for taking a look. Actually I identified some more to improve for this command. I will add a bit more change and a unit test and then let you know.

jiacheliu3 commented:
TODO: 
1. Make this work in HA
2. Add unit tests

jiacheliu3 commented on shell/src/main/java/alluxio/cli/LogLevel.java:
```diff
@@ -56,15 +59,17 @@
+  public static final String LOG_LEVEL = "logLevel";
+  public static final String ROLE_WORKERS = "workers";
+  public static final String ROLE_MASTER = "master";
+  public static final String ROLE_MASTERS = "masters";
+  public static final String ROLE_WORKER = "worker";
+  public static final String ROLE_JOB_MASTER = "job_master";
+  public static final String ROLE_JOB_MASTERS = "job_masters";
```
`masters` and `job_masters` are not valid options. Instead of rejecting them, we give a warning and translate it to the correct one.

jiacheliu3 commented on shell/src/main/java/alluxio/cli/LogLevel.java:
```diff
@@ -267,25 +340,44 @@ public static void main(String[] args) {
+  /**
+   * Object that represents a REST endpoint that logLevel sends HTTP request to.
+   * */
+  public static final class TargetInfo {
```
Change to public for better test visibility

jiacheliu3 commented on shell/src/main/java/alluxio/cli/LogLevel.java:
```diff
@@ -141,32 +154,68 @@ public static void logLevel(String[] args, AlluxioConfiguration alluxioConf)
+          jobClient = JobMasterClient.Factory.create(JobMasterClientContext
+                  .newBuilder(clientContext).build());
+        }
+        String jobMasterHost = jobClient.getAddress().getHostName();
```
Find primary

jiacheliu3 commented on shell/src/main/java/alluxio/cli/LogLevel.java:
```diff
@@ -141,32 +154,68 @@ public static void logLevel(String[] args, AlluxioConfiguration alluxioConf)
+        if (fsContext == null) {
+          fsContext = FileSystemContext.create(clientContext);
+        }
+        String masterHost = fsContext.getMasterAddress().getHostName();
```
Find primary

jiacheliu3 commented on shell/src/main/java/alluxio/cli/LogLevel.java:
```diff
@@ -141,32 +154,68 @@ public static void logLevel(String[] args, AlluxioConfiguration alluxioConf)
   private static List<TargetInfo> getTargetInfos(String[] targets, AlluxioConfiguration conf)
       throws IOException {
+    // Trim the elements
+    Set<String> targetSet = Arrays.stream(targets).map(String::trim).collect(Collectors.toSet());
```
Be more tolerant by trimming

LuQQiu commented on docs/en/operation/User-CLI.md:
```diff
@@ -235,16 +235,19 @@ where:
 
 > Note: This command requires the Alluxio cluster to be running.
+> You are not able to set the logger level on the standby masters.
+> The standby masters/job masters do not have a running web server.
```
Does our primary job master have a web server?
or are you mentioning the running RPC server?

jiacheliu3 replied:
Yes the job master has similar web server logic as the master process. It does not have a web UI but accepts a few HTTP calls.  See `AlluxioJobMasterRestServiceHandler`.

LuQQiu commented on shell/src/main/java/alluxio/cli/LogLevel.java:
```diff
@@ -141,32 +154,68 @@ public static void logLevel(String[] args, AlluxioConfiguration alluxioConf)
+
+    // Allow plural form for the master/job_master and print a notice
+    if (targetSet.contains(ROLE_MASTERS) || targetSet.contains(ROLE_JOB_MASTERS)) {
+      System.out.println("The logLevel command will only take effect on the primary master, "
```
primary master node may not be the primary job master node, can we give more clear warning message here for different roles?

LuQQiu commented on shell/src/main/java/alluxio/cli/LogLevel.java:
```diff
@@ -227,7 +280,7 @@ private static String parseOptLevel(CommandLine cmd) {
 
   private static void setLogLevel(final TargetInfo targetInfo, String logName, String level,
-                                  AlluxioConfiguration alluxioConf)
+                                  AlluxioConfiguration conf)
```
indent

LuQQiu commented on shell/src/main/java/alluxio/cli/LogLevel.java:
```diff
@@ -246,6 +299,26 @@ private static void setLogLevel(final TargetInfo targetInfo, String logName, Str
+      return ROLE_JOB_WORKER;
+    } else {
+      throw new IllegalArgumentException(
+              String.format("Unrecognized port in %s. " + "Please make sure the port is in %s",
```
indent

LuQQiu reviewed (commented):
Thanks for making the logging change much easier!
From the high level, as an user, i may be confused on the first several line and thinks they are not needed? The only information they add to the future lines is the master we modified logs on is the primary master, the job master we modified logs on is the primary job master. Those information also be shown in the docs and command --help ?
```
$ bin/alluxio logLevel --logName=alluxio.master.job.plan --level=DEBUG
Primary job master: **AlluxioSandbox-log1-masters-1**:20002[job_master]
Job worker: AlluxioSandbox-log1-workers-2:30003[job_worker]
Job worker: AlluxioSandbox-log1-workers-1:30003[job_worker]
Worker: AlluxioSandbox-log1-workers-2:30000[worker]
Worker: AlluxioSandbox-log1-workers-1:30000[worker]
Primary master: **AlluxioSandbox-log1-masters-2**:19999[master]
...
```

jiacheliu3 commented:
> Thanks for making the logging change much easier!
> From the high level, as an user, i may be confused on the first several line and thinks they are not needed? The only information they add to the future lines is the master we modified logs on is the primary master, the job master we modified logs on is the primary job master. Those information also be shown in the docs and command --help ?
> 
> ```
> $ bin/alluxio logLevel --logName=alluxio.master.job.plan --level=DEBUG
> Primary job master: **AlluxioSandbox-log1-masters-1**:20002[job_master]
> Job worker: AlluxioSandbox-log1-workers-2:30003[job_worker]
> Job worker: AlluxioSandbox-log1-workers-1:30003[job_worker]
> Worker: AlluxioSandbox-log1-workers-2:30000[worker]
> Worker: AlluxioSandbox-log1-workers-1:30000[worker]
> Primary master: **AlluxioSandbox-log1-masters-2**:19999[master]
> ...
> ```

Good question. Yea these output are probably redundant with the lines below. I removed them. 

Regarding the doc and `--help` the only update in the usage is in `User-CLI.md`, as you see in this PR.

jiacheliu3 commented:
@LuQQiu PTAL, if the change looks good I will merge it.

LuQQiu commented on shell/src/main/java/alluxio/cli/LogLevel.java:
```diff
@@ -309,13 +309,13 @@ private static String inferRoleFromPort(int port, AlluxioConfiguration conf) {
-                        PropertyKey.JOB_MASTER_WEB_PORT, PropertyKey.JOB_WORKER_WEB_PORT
-                      })));
+      throw new IllegalArgumentException(String.format(
+              "Unrecognized port in %s. " + "Please make sure the port is in %s",
```
indent

LuQQiu reviewed (approved):
LGTM!

jiacheliu3 commented:
alluxio-bot, merge this please
